Firestore/core/src/firebase/firestore/api/document_reference.mm

```diff
@@ -127,12 +127,16 @@ ListenOptions options(
       /*include_document_metadata_changes=*/true,
       /*wait_for_sync_when_online=*/true);
 
+  // Create an empty ListenerRegistration captured in the closure below and
+  // then, once the listen has started, assign to it and signal the listener
+  // that we succeeded.
+  //
   // TODO(varconst): replace with a synchronization primitive that doesn't
   // require libdispatch. See
   // https://github.com/firebase/firebase-ios-sdk/blob/3ccbdcdc65c93c4621c045c3c6d15de9dcefa23f/Firestore/Source/Core/FSTFirestoreClient.mm#L161
   // for an example.
   dispatch_semaphore_t registered = dispatch_semaphore_create(0);
-  auto listener_registration = std::make_shared<id<FIRListenerRegistration>>();
+  auto listener_registration = std::make_shared<ListenerRegistration>();
```

Comment:
Very optional: I think you can just create a value now and copy it into the lambda?

Reply:
I can't copy into the lambda because the creator of the lambda needs to actually populate this value. There needs to be some shared storage between them. Note that this becomes a `std::promise` in the next PR which makes this easier to understand.